src/Raven.Server/Documents/Sharding/Queries/AbstractShardedQueryProcessor.cs

```diff
@@ -213,24 +218,81 @@ private BlittableJsonReaderObject RewriteQueryIfNeeded(BlittableJsonReaderObject
     {
         var rewriteForPaging = Query.Offset is > 0;
         var rewriteForProjectionFromMapReduceIndex = true;
-        var rewriteForProjectionFromAutoMapReduceIndex = true;
+        var rewriteForProjectionFromAutoMapReduceQuery = true;
 
         var query = Query.Metadata.Query;
         var isProjectionQuery = query.Select?.Count > 0 ||
-                 query.SelectFunctionBody.FunctionText != null;
+                                query.SelectFunctionBody.FunctionText != null;
+
+        var rewriteForFilterForMapReduceIndexOrAutoMapReduceQuery = false;
+        var rewriteForLimitWithOrderByInMapReduce = false;
+        List<string> setOrderByFieldsForMapReduce = null;
+
+        if (IsMapReduceIndex || IsAutoMapReduceQuery)
+        {
+            rewriteForFilterForMapReduceIndexOrAutoMapReduceQuery = query.Filter != null;
+
+            if (query.Limit != null)
+            {
+                List<string> groupByFields;
+
+                if (IsAutoMapReduceQuery)
+                {
+                    groupByFields = Query.Metadata.GroupBy.Select(x => x.Name.Value).ToList();
+                }
+                else
+                {
+                    Debug.Assert(IsMapReduceIndex);
+                    var index = RequestHandler.DatabaseContext.Indexes.GetIndex(Query.Metadata.IndexName);
+                    if (index == null)
+                        IndexDoesNotExistException.ThrowFor(Query.Metadata.IndexName);
+
+                    if (index.Type.IsAutoMapReduce())
+                    {
+                        var autoMapReduceIndexDefinition = (AutoMapReduceIndexDefinition)index.Definition;
+                        groupByFields = autoMapReduceIndexDefinition.GroupByFields.Keys.ToList();
+                    }
+                    else
+                    {
+                        if (index.Type.IsStaticMapReduce() == false)
+                            throw new InvalidOperationException($"Index '{Query.Metadata.IndexName}' is not a map-reduce index");
+
+                        var compiled = ((StaticIndexInformationHolder)index).Compiled;
+                        groupByFields = compiled.GroupByFields.Select(x => x.Name).ToList();
+                    }
+                }
+
+                if (query.OrderBy != null)
+                {
+                    // we have a limit and order by fields
+                    // we need to get all results if the order by isn't done on the group by fields
+
+                    var orderByFields = Query.Metadata.OrderBy.Select(x => x.Name.Value).ToList();
```

Comment:
All of those `ToList()` can be expensive, can we move that to the query metadata so it only happens once?

Reply:
done